retryrpc/client.go

```diff
@@ -391,6 +423,7 @@ func (client *Client) dial() (err error) {
 		client.connection.tlsConn.Close()
 		client.connection.tlsConn = nil
 	}
+
 	client.connection.tlsConn = tlsConn
 	client.connection.state = CONNECTED
 	client.connection.genNum++
```

Comment:
The 3 lines setting the state to CONNECTED should move moved after the check for error from *sendMyInfo()*

If *sendMyInfo()* fails, we should NOT change the connection state since it could be in RETRANSMITTING and would open a subtle race condition where another goroutine thinks it needs to initiate retransmitting in parallel with this goroutine.

I think there is a bug in the existing code which we have not hit yet and I think by moving the setting of CONNECTED to after the call to sendMyInfo() we avoid the problem.